iroh-relay/src/protos/relay.rs

```diff
@@ -568,385 +321,301 @@ impl Frame {
+    ///
+    /// Specifically meant for being put into a binary websocket message frame.
+    pub(crate) fn write_to<O: BufMut>(&self, mut dst: O) -> O {
+        dst = self.typ().write_to(dst);
+        match self {
+            Self::Datagrams {
+                dst_node_id,
+                datagrams,
+            } => {
+                dst.put(dst_node_id.as_ref());
+                dst = datagrams.write_to(dst);
             }
-
-            // Can't use the `Buf::get_*` APIs, as that advances the buffer.
-            let Some(frame_type) = src.first().map(|b| FrameType::from(*b)) else {
-                return Ok(None); // Not enough bytes
-            };
-            let Some(frame_len) = src
-                .get(1..5)
-                .and_then(|s| TryInto::<[u8; 4]>::try_into(s).ok())
-                .map(u32::from_be_bytes)
-                .map(|l| l as usize)
-            else {
-                return Ok(None); // Not enough bytes
-            };
-
-            if frame_len > MAX_FRAME_SIZE {
-                return Err(FrameTooLargeSnafu { frame_len }.build());
+            Self::Ping(data) => {
+                dst.put(&data[..]);
             }
-
-            if src.len() < HEADER_LEN + frame_len {
-                // Optimization: prereserve the buffer space
-                src.reserve(HEADER_LEN + frame_len - src.len());
-
-                return Ok(None);
+            Self::Pong(data) => {
+                dst.put(&data[..]);
             }
-
-            // advance the header
-            src.advance(HEADER_LEN);
-
-            let content = src.split_to(frame_len).freeze();
-            let frame = Frame::from_bytes(frame_type, content, &self.cache)?;
-
-            Ok(Some(frame))
         }
+        dst
     }
 
-    impl Encoder<Frame> for RelayCodec {
-        type Error = std::io::Error;
-
-        fn encode(&mut self, frame: Frame, dst: &mut BytesMut) -> Result<(), Self::Error> {
-            let frame_len: usize = frame.len();
-            if frame_len > MAX_FRAME_SIZE {
-                return Err(std::io::Error::new(
-                    std::io::ErrorKind::InvalidData,
-                    format!("Frame of length {frame_len} is too large."),
-                ));
-            }
-
-            let frame_len_u32 = u32::try_from(frame_len).expect("just checked");
-
-            dst.reserve(HEADER_LEN + frame_len);
-            dst.put_u8(frame.typ().into());
-            dst.put_u32(frame_len_u32);
-            frame.write_to(dst);
-
-            Ok(())
-        }
-    }
-}
+    /// Tries to decode a frame received over websockets.
+    ///
+    /// Specifically, bytes received from a binary websocket message frame.
+    #[allow(clippy::result_large_err)]
+    #[cfg(feature = "server")]
+    pub(crate) fn from_bytes(bytes: Bytes, cache: &KeyCache) -> Result<Self, Error> {
+        let (frame_type, content) = FrameType::from_bytes(bytes).context(InvalidFrameSnafu)?;
```

Comment:
IIUC this invalid (or rather, unknown?) frame gets propagated all the way the the client Stream interface.  So it is clients themselves that decide what to do with this.  In our case the ActiveRelayActor stops when it receives an unknown frame.

However, if we allow skipping over any unknown frames - we could do that right here - then we can gracefully extend the protocol by adding new frame types.  Older clients would ignore them and we'd only need to be careful to not keep sending them if the client ignore them.

It would also simplify a lot of downstream use as the stream just becomes `Item = Option<Frame>` rather than the current `Item = Option<Result<Frame, ()>>`.

Reply:
Thanks for noting this. I changed this so that there's no more `FrameType::Unknown`.

I'm still keeping the `Result` though. E.g. in the `handshake` protocol it's nice to not skip unknown frames, since any unknown frame is immediately an error.